### PR TITLE
Port CA1033: Interface methods should be callable by child types (htt…

### PR DESCRIPTION
**Pull request by mavasani**
…ps://msdn.microsoft.com/library/ms182153.aspx) to FxCop analyzers project.

### CODE REVIEW

**Comment by mavasani** [comment]
@tmeschter @pharring @srivatsn @heejaechang @shyamnamboodiripad @ManishJayaswal can you please review?

File: src/Diagnostics/FxCop/Core/AbstractCodeBlockAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.FxCopAnalyzers
+{
+    public abstract class AbstractCodeBlockAnalyzer : DiagnosticAnalyzer
+    {
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
```
Comment (srivatsn) [inline src/Diagnostics/FxCop/Core/AbstractCodeBlockAnalyzer.cs]: Do we really need these Abstract types (this and the SyntaxNode one)? I don't think they add much value - they reduce a couple of lines of typing but I think they just add complexity to the analyzer in that it's all not in one place. Seeing the Initialize method, in my opinion, is a much clearer way of understanding what the analyzer is doing.

Comment (mavasani) [inline src/Diagnostics/FxCop/Core/AbstractCodeBlockAnalyzer.cs]: @srivatsn Probably that's just an alternate preference :). I have removed the abstract code block analyzer type I added here.

For me, an analyzer (especially simple stateless ones) is more cleaner if it is declared as one via the type hierarchy (named type analyzer, syntax analyzer, etc.). What I prefer to read is: okay, I am a code block analyzer and I need an AnalyzeCodeBlock method. The option you suggested is that all analyzers are DiagnosticAnalyzer and you need to dig through the Initialize method and registered actions to figure out what it is doing. Later option is good in the sense that everything about the analyzer implementation is in one place, the former option is good in the sense that everything about "my analyzer" implementation is declaratively understood by looking at just the base types and overridden methods :)

**Comment by mavasani** [comment]
@dotnet-bot test this please.

**Comment by mavasani** [comment]
I have added a code fixer for CA1033, and corresponding unit tests. Can you please review?

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**